// skyflow-node/lib/vault-api/utils/common/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUREJS_TYPES = exports.MessageType = exports.LogLevel = exports.RequestMethod = exports.RedactionType = void 0;
var RedactionType;
(function (RedactionType) {
    RedactionType["DEFAULT"] = "DEFAULT";
    RedactionType["PLAIN_TEXT"] = "PLAIN_TEXT";
    RedactionType["MASKED"] = "MASKED";
    RedactionType["REDACTED"] = "REDACTED";
})(RedactionType = exports.RedactionType || (exports.RedactionType = {}));
var RequestMethod;
(function (RequestMethod) {
    RequestMethod["GET"] = "GET";
    RequestMethod["POST"] = "POST";
    RequestMethod["PUT"] = "PUT";
    RequestMethod["PATCH"] = "PATCH";
    RequestMethod["DELETE"] = "DELETE";
})(RequestMethod = exports.RequestMethod || (exports.RequestMethod = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel["WARN"] = "WARN";
    LogLevel["INFO"] = "INFO";
    LogLevel["DEBUG"] = "DEBUG";
    LogLevel["ERROR"] = "ERROR";
    LogLevel["NONE"] = "NONE";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
var MessageType;
(function (MessageType) {
    MessageType["LOG"] = "LOG";
    MessageType["WARN"] = "WARN";
    MessageType["ERROR"] = "ERROR";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
exports.PUREJS_TYPES = {
    INSERT: 'INSERT',
    DETOKENIZE: 'DETOKENIZE',
    GET_BY_SKYFLOWID: 'GET_BY_SKYFLOWID',
    INVOKE_CONNECTION: 'INVOKE_CONNECTION',
};
